src/Storages/StorageBuffer.cpp

```diff
@@ -631,12 +618,11 @@ bool StorageBuffer::mayBenefitFromIndexForIn(
     if (!destination_id)
         return false;
 
-    auto destination = DatabaseCatalog::instance().getTable(destination_id, query_context);
-
-    if (destination.get() == this)
-        throw Exception("Destination table is myself. Read will cause infinite loop.", ErrorCodes::INFINITE_LOOP);
+    auto destination = getDestinationTable();
+    if (destination)
+        return destination->mayBenefitFromIndexForIn(left_in_operand, query_context, destination->getInMemoryMetadataPtr());
```

Comment:
@alexey-milovidov could you also modify this in your PR, before it does not check that the table still exists